pinephone/vibrator/FFDevice.cpp

```diff
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (C) 2019 The Android Open-Source Project
+ * Copyright (C) 2021 GloDroid project
+ */
+
+#include "vibrator-impl/FFDevice.h"
+
+#include <android-base/logging.h>
+
+typedef bool (*dirfunc_t)(const char *path, void *param);
+static bool recursive_dir_observer(const char *path, dirfunc_t func,
+                                   void *param);
+
+namespace aidl {
+namespace android {
+namespace hardware {
+namespace vibrator {
+
+std::unique_ptr<FFDevice> FFDevice::create(const char *input_path) {
+    std::string event_path;
+    auto observe_func = [](const char *path, void *param) -> bool {
+        std::string *event_path = static_cast<std::string *>(param);
+        std::string p = path;
+        if (p.find("event") != std::string::npos) {
+            *event_path = p;
+            return true;
+        }
+        return false;
+    };
+    if (recursive_dir_observer(input_path, observe_func, &event_path)) {
+        event_path = "/dev/input/" + event_path;
+        auto device = std::unique_ptr<FFDevice>(new FFDevice());
+        device->event_fd = open(event_path.c_str(), O_RDWR);
+        if (device->event_fd == -1) {
+            LOG(ERROR) << "FFDevice: error opening event file";
+            return nullptr;
+        }
+        return device;
+    }
+    LOG(ERROR) << "FFDevice: event path not found";
+    return nullptr;
+}
+
+FFDevice::~FFDevice() {
+    if (event_fd >= 0) {
+        close(event_fd);
+    }
+}
+
+void FFDevice::vibrate(int duration_ms) {
+    if (last_effect_id != -1) {
+        off();
+    }
+
+    /* Prepare effect */
+    ff_effect effect = {.type = FF_RUMBLE,
+                        .u.rumble.strong_magnitude = 65535,
```

Comment:
Why 65535? Do you want specifically this value, or do you want just `std::numeric_limits<uint16_t>::max`? If later, then consider replacing this value with constant, if former, then put a comment on why 65535.